### PR TITLE
Implement functional :host pseudo-class in Matches

### PR DESCRIPTION
**Pull request by 2biazdk**
Also add negative test cases.

### CODE REVIEW

File: packages/alfa-dom/src/matches.ts [hunk 1]
```diff
@@ -430,12 +430,24 @@ function matchesPseudoClass(
...
+        matches(element, context, selector.value, options, root);
+
       return (
-        root === selector &&
-        options.treeContext !== undefined &&
         isShadowRoot(options.treeContext) &&
         getParentElement(options.treeContext, context, { composed: true }) ===
-          element
+          element &&
+        isHostMatched
```
Comment (NiclasHedam) [inline packages/alfa-dom/src/matches.ts]: There is something funky about these indents, isn't there?

Comment (2biazdk) [reply packages/alfa-dom/src/matches.ts]: It. Is. Impossible to fix. 😭

Comment (kasperisager) [reply packages/alfa-dom/src/matches.ts]: You can pull out `getParentElement(options.treeContext, context, { composed: true })` to a variable and reference that instead. [Prettier](https://github.com/prettier/prettier) is pretty particular on indents 🙈

File: packages/alfa-dom/src/matches.ts [hunk 2]
```diff
@@ -430,12 +430,24 @@ function matchesPseudoClass(
 
     // https://drafts.csswg.org/css-scoping/#host-selector
     case "host":
+      // Do not allow prefix (e.g. "div:host")
+      if (root !== selector) {
+        return false;
+      }
+
+      if (options.treeContext === undefined) {
```
Comment (kasperisager) [inline packages/alfa-dom/src/matches.ts]: This should also check for `isShadowRoot(options.treeContext)` so as to bail out early if the tree context isn't a shadow root.

File: packages/alfa-dom/src/matches.ts [hunk 3]
```diff
@@ -430,12 +430,24 @@ function matchesPseudoClass(
...
+        return false;
+      }
+
+      const host = getParentElement(options.treeContext, context, {
+        composed: true
+      });
+      const isHostMatched =
+        selector.value === null ||
+        matches(element, context, selector.value, options, root);
+
```
Comment (kasperisager) [inline packages/alfa-dom/src/matches.ts]: Doing this...

```ts
if (host !== element) {
  return false;
}

return selector.value === null || matches(element, context, selector.value, options, root);
```

...will ensure that we don't attempt to do selector matching unless we absolutely have to.

**Pull request merged**